src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py

```diff
@@ -420,6 +420,32 @@ def __copy__(self) -> "SnowflakePlan":
                 placeholder_query=self.placeholder_query,
             )
 
+    def __deepcopy__(self, memodict={}) -> "SnowflakePlan":  # noqa: B006
+        copied_plan = SnowflakePlan(
+            queries=copy.deepcopy(self.queries) if self.queries else [],
```

Comment:
SnowflakePlan also has some cached_properties like `attributes` which run a describe query once and then reuse the result from subsequent calls. During this deepcopy, we do not borrow those results over. I'm wondering if that can cause any issues down the line.

Reply:
The attribute result is get from the schema call, during the copy, we make sure the schema queries are copied correctly, so the worst case is that a if the attribute call is called on the copied plan, a describing call will be triggered.  In our test, we do check the schema query is copied correctly. 
Theoretically, all those properties should be able to be reconstructed correctly if all fields of the class is copied correctly. 
We can also make a copy of all the cached_property if exists, which is probably not necessary now, since the deep copy is currently only used by our new compilation stage, and none of our optimization step should require the attribute property.